vmsyscall/src/memory_map.rs

```diff
@@ -0,0 +1,317 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! The `MemoryMap` describes the virtual machine address space.
+//!
+//! Initially copied from:
+//! https://github.com/rust-osdev/bootloader/blob/90f5b8910d146d6d489b70a6341d778253663cfa/src/bootinfo/memory_map.rs
+
+use core::cmp::Ordering;
+use core::fmt;
+use core::ops::{Deref, DerefMut};
+
+pub(crate) const PAGE_SIZE: u64 = 4096;
+
+const MAX_MEMORY_MAP_SIZE: usize = 64;
+
+/// A map of the physical memory regions of the underlying machine.
+#[derive(Clone)]
+#[repr(C)]
+pub struct MemoryMap {
+    entries: [MemoryRegion; MAX_MEMORY_MAP_SIZE],
+    // u64 instead of usize so that the structure layout is platform
+    // independent
+    next_entry_index: u64,
+}
+
+impl MemoryMap {
+    /// Produce an empty `MemoryMap`.
+    pub const fn new() -> Self {
+        MemoryMap {
+            entries: [MemoryRegion::empty(); MAX_MEMORY_MAP_SIZE],
+            next_entry_index: 0,
+        }
+    }
+
+    /// Mark a region as usable.
+    pub fn set_region_type_usable(&mut self, region_type: MemoryRegionType) {
+        self.iter_mut().for_each(|r| {
+            if r.region_type == region_type {
+                r.region_type = MemoryRegionType::Usable
+            }
+        });
+    }
+
+    /// Add a region to the `MemoryMap`.
+    pub fn add_region(&mut self, region: MemoryRegion) {
+        if let Err(()) = self.entries.iter_mut().try_for_each(|last_region| {
+            if last_region.region_type == region.region_type
+                && last_region.range.end_frame_number >= region.range.start_frame_number
+                && last_region.range.end_frame_number <= region.range.end_frame_number
+            {
+                last_region.range.end_frame_number = region.range.end_frame_number;
+                return Err(());
+            }
+            Ok(())
+        }) {
+            return;
+        };
+
+        assert!(
+            self.next_entry_index() < MAX_MEMORY_MAP_SIZE,
+            "too many memory regions in memory map"
+        );
+
+        self.entries[self.next_entry_index()] = region;
+        self.next_entry_index += 1;
+        self.sort();
+    }
+
+    /// Mark a region as allocated.
+    pub fn mark_allocated_region(&mut self, region: MemoryRegion) {
+        let mut region = region;
+        for r in self.iter_mut() {
+            // New region inside region of same type
+            if r.region_type == region.region_type
+                && r.range.start_frame_number <= region.range.start_frame_number
+                && r.range.end_frame_number >= region.range.end_frame_number
```

Comment:
This pattern or similar occurs frequently. Could we make this an operation on the `MemoryRegion` struct?